# Домашнее задание 2 (5 баллов).

Все задания ниже имеют равный вес (5/16).

In [ ]:
# импортируем пандас
import pandas as pd

#### Описание данных

В папке Dat (https://github.com/hse-ds/iad-intro-ds/blob/master/2023/homeworks/Data.zip) находится информация о студентах. Всего 10 групп студентов. Файлы делятся на две категории:
    * Students_info_i - информация о студентах из группы i
    * Students_marks_i - оценки студентов из группы i за экзамены

### Одно из важных достоинств pandas $-$ это удобные методы реляционного взаимодействия с данными, аналогичные, например, возможностям SQL для слияния и конкатенации таблиц: merge, join, concat. Наличие готовых методов позволяет не реализовывать самостоятельно поэлементную обработку данных и оперировать сразу целыми таблицами данных.

Подробно об этих методах посмотрите тут: https://www.kaggle.com/residentmario/renaming-and-combining#Combining

#### Задание 1. Соберите всю информацию о студентах в одну таблицу df. В получившейся таблице должна быть информация и оценки всех студентов из всех групп. Напечатайте несколько строк таблицы для демонстрации результата.

In [ ]:
# устанавливаем библиотеку wget, чтобы загрузить файлы
!pip install wget

  Preparing metadata (setup.py) ... done
  Created wheel for wget: filename=wget-3.2-py3-none-any.whl size=9655 sha256=337060697af5382b5edeccd9fdbb20fab930c8d5a9c47089576424b1b42e38af
  Stored in directory: /root/.cache/pip/wheels/8b/f1/7f/5c94f0a7a505ca1c81cd1d9208ae2064675d97582078e6c769
Successfully built wget


In [ ]:
# импортируем wget
import wget

# скачиваем зип-файл по ссылке
url = 'https://github.com/hse-ds/iad-intro-ds/raw/master/2023/homeworks/Data.zip'
filename = wget.download(url)

In [ ]:
# импортируем библиотеку zipfile, чтобы разархивировать скачанный зип-файл
import zipfile

# разархивируем все файлы
with zipfile.ZipFile(filename, 'r') as zip_ref:
    zip_ref.extractall()

In [ ]:
# импортируем библиотеку os, чтобы открыть все файлы в директории
import os

# все файлы были разархивированы в директории, у которой название такое же,
# как у зип-файла, но без .zip
path = filename.replace('.zip', '')

# в этих двух списках будут хранится таблички пандас, потом соединим их
students_info = []
students_marks = []
# для каждого файла в директории (в папке)
for fi in sorted(os.listdir(path)):
    # если в названии файла есть 'Students_info_'
    if 'Students_info_' in fi:
        # читаем csv файл, делаем его таблицей в пандасе и добавляем в список
        students_info_i = pd.read_csv(f'{path}/{fi}')
        students_info.append(students_info_i)
    # аналогично, если в названии есть 'Students_marks_'
    elif 'Students_marks_' in fi:
        # делаем файл таблицей в пандасе и добавляем в список
        students_marks_i = pd.read_csv(f'{path}/{fi}')
        students_marks.append(students_marks_i)
# создаем общую таблицу всех студентов
info = pd.concat(students_info)
# создаем общую таблицу всех оценок
marks = pd.concat(students_marks)
# соединяем получившиеся таблицы по индексу, outer join'ом
# (чтобы нулевые значения тоже сохранились)
df = pd.merge(info, marks, on="index", how="outer")
# выводим таблицу (5 произвольных строк)
df.sample(5)

index  gender race/ethnicity parental level of education         lunch  \
902    902  female        group A                 high school  free/reduced   
243    243    male        group E                some college      standard   
564    564    male        group B           bachelor's degree  free/reduced   
671    671    male        group D          associate's degree  free/reduced   
361    361    male        group B            some high school      standard   

    test preparation course    group  math score  reading score  writing score  
902               completed  group10          34             48             41  
243                    none   group3          59             51             43  
564                    none   group6          48             51             46  
671                    none   group7          52             57             50  
361               completed   group4          85             84             78

#### Задание 2. Удалите столбец index у полученной таблицы. Напечатайте первые 10 строк таблицы.

In [ ]:
# удаляем столбец с названием 'index'
df = df[df.columns.drop("index")]
# выводим первые 10 строк
df.head(10)

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   
5  female        group B          associate's degree      standard   
6  female        group B                some college      standard   
7    male        group B                some college  free/reduced   
8    male        group D                 high school  free/reduced   
9  female        group B                 high school  free/reduced   

  test preparation course   group  math score  reading score  writing score  
0                    none  group1          72             72             74  
1               completed  group1          69             90             88  
2                    none  group1          90             95             93  
3                    none  group1          47             57             44  
4                    none  group1          76             78             75  
5                    none  group1          71             83             78  
6               completed  group1          88             95             92  
7                    none  group1          40             43             39  
8               completed  group1          64             64             67  
9                    none  group1          38             60             50

#### Задание 3. Выведите на экран размеры полученной таблицы

In [ ]:
# выводим размер с помощью shape
# 1000 строк, 9 столбцов
df.shape

(1000, 9)

#### Задание 4. Выведите на экран статистические характеристики числовых столбцов таблицы (минимум, максимум, среднее значение, стандартное отклонение)

In [ ]:
# с помощью describe выводим статистические характеристики числовых столбцов
df.describe()

math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

#### Задание 5. Проверьте, есть ли в таблице пропущенные значения

In [ ]:
# проверяем количество пропущенных значений по каждому столбцу
# их нет ни в одном столбце
df.isna().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
group                          0
math score                     0
reading score                  0
writing score                  0
dtype: int64

#### Задание 6. Выведите на экран средние баллы студентов по каждому предмету (math, reading, writing)

In [ ]:
# выводим среднее по столбцам с каждым из предметов
df[["math score","reading score", "writing score"]].mean()

math score       66.089
reading score    69.169
writing score    68.054
dtype: float64

**Задание 7. Как зависят оценки от того, проходил ли студент курс для подготовки к сдаче экзамена (test preparation course)? Выведите на экран для каждого предмета в отдельности средний балл студентов, проходивших курс для подготовки к экзамену и не проходивших курс.**

In [ ]:
# группируем данные по значениям в столбце test prepatation course
# в значениях будет указано среднее по столбцам с каждым предметом
# в целом средний балл по всем предметам тех, кто прошел курс, выше
df.groupby("test preparation course")[["math score","reading score", "writing score"]].mean()

math score  reading score  writing score
test preparation course                                          
completed                 69.695531      73.893855      74.418994
none                      64.077882      66.534268      64.504673

**Задание 8. Выведите на экран все различные значения из столбца lunch.**

In [ ]:
# выводим при помощи unique различные значения столбца
print(df["lunch"].unique())

['standard' 'free/reduced']


**Задание 9. Переименуйте колонку "parental level of education" в "education", а "test preparation course" в "test preparation" с помощью метода pandas rename**
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rename.html

In [ ]:
# меняем названия колонок в таблице, меняем саму таблицу (чтобы изменения сохранились)
df = df.rename(columns={"parental level of education": "education",
                        "test preparation course": "test preparation"})
df

gender race/ethnicity           education         lunch test preparation  \
0    female        group B   bachelor's degree      standard             none   
1    female        group C        some college      standard        completed   
2    female        group B     master's degree      standard             none   
3      male        group A  associate's degree  free/reduced             none   
4      male        group C        some college      standard             none   
..      ...            ...                 ...           ...              ...   
995  female        group E     master's degree      standard        completed   
996    male        group C         high school  free/reduced             none   
997  female        group C         high school  free/reduced        completed   
998  female        group D        some college      standard        completed   
999  female        group D        some college  free/reduced             none   

       group  math score  reading score  writing score  
0     group1          72             72             74  
1     group1          69             90             88  
2     group1          90             95             93  
3     group1          47             57             44  
4     group1          76             78             75  
..       ...         ...            ...            ...  
995  group10          88             99             95  
996  group10          62             55             55  
997  group10          59             71             65  
998  group10          68             78             77  
999  group10          77             86             86  

[1000 rows x 9 columns]

**Зафиксируем минимальный балл для сдачи экзамена**

In [ ]:
# указываем passmark
passmark = 50

**Задание 10. Ответьте на вопросы:**
    * Какая доля студентов сдала экзамен по математике (passmark >= 50)?
    * Какая доля студентов, проходивших курс подготовки к экзамену, сдала экзамен по математике?
    * Какая доля женщин, не проходивших курс подготовки к экзамену, не сдала экзамен по математике?

 Какая доля студентов сдала экзамен по математике (passmark >= 50)?

In [ ]:
# посчитаем количество учеников, сдававших экзамен по математике
total = len(df['math score'])
# посчитаем количество учеников, у которых балл за экзамен больше или равен passmark
passed = len(df[df['math score'] >= passmark])
# доля сдавших экзамен по математике -- отношение количества тех, кто сдал, на количество сдававших
print(passed / total)

0.865


Какая доля студентов, проходивших курс подготовки к экзамену, сдала экзамен по математике?

In [ ]:
# здесь будут все, кто готовился к экзамену
prepared = df[df["test preparation"] == "completed"]
# здесь среди тех, кто готовился к экзамену, те, кто сдал его (>= passmark)
passed_prepared = prepared[prepared["math score"] >= passmark]
# делим количество сдавших и подготовленных на количество подготовленных
print(len(passed_prepared) / len(prepared))

0.9217877094972067


Какая доля женщин, не проходивших курс подготовки к экзамену, не сдала экзамен по математике?

In [ ]:
# здесь все женщины, которые не готовились к экзамену
not_prepared = df[(df["test preparation"] == "none") & (df["gender"] == "female")]
# здесь те женщины, которые не готовились и не сдали
w_np = not_prepared[not_prepared["math score"] < passmark]
# находим отношение не сдавших и не подготовленных к экзамену женщин к не подготовленным к экзамену женщинам
print(len(w_np) / len(not_prepared))

0.20958083832335328


**Задание 11. С помощью groupby выполните задания ниже. Также выведите время выполнения каждого из заданий.**
    * Для каждой этнической группы выведите средний балл за экзамен по чтению
    * Для каждого уровня образования выведите минимальный балл за экзамен по письму

Для каждой этнической группы выведите средний балл за экзамен по чтению


In [ ]:
# чтобы вывести время
%%time
# группируем по этническим группам и выводим средний балл по чтению для каждой
df.groupby("race/ethnicity")["reading score"].mean()

CPU times: user 3.24 ms, sys: 0 ns, total: 3.24 ms
Wall time: 3.26 ms


race/ethnicity
group A    64.674157
group B    67.352632
group C    69.103448
group D    70.030534
group E    73.028571
Name: reading score, dtype: float64

Для каждого уровня образования выведите минимальный балл за экзамен по письму

In [ ]:
# чтобы вывести время работы
%%time
# группируем по уровню образования и выводим минимальный балл по письму для каждого
df.groupby("education")["writing score"].apply(min)

CPU times: user 5.55 ms, sys: 814 µs, total: 6.36 ms
Wall time: 14.9 ms


education
associate's degree    35
bachelor's degree     38
high school           15
master's degree       46
some college          19
some high school      10
Name: writing score, dtype: int64

**Задание 12. Выполните задание 11 с помощью циклов. Сравните время выполнения.**

Для каждой этнической группы выведите средний балл за экзамен по чтению

In [ ]:
# чтобы засечь время
%%time
# создаем словарь с национальностями, куда будем записывать баллы по чтению
ethnicities = {}
# для каждой строчки в таблцие
for i in range(len(df['race/ethnicity'])):
    # если национальность еще не было в словаре
    if df['race/ethnicity'][i] not in ethnicities:
        # создаем для нее пустой словарь
        ethnicities[df['race/ethnicity'][i]] = []
        # затем добавляем в список балл по чтению для каждой строки
        ethnicities[df['race/ethnicity'][i]].append(df['reading score'][i])
    # если национальность была в словаре
    else:
        # добавляем в список балл по чтению
        ethnicities[df['race/ethnicity'][i]].append(df['reading score'][i])
# для каждого ключа в словаре (каждой национальности)
for key in ethnicities:
    # присваиваем вместо списка среднее его значение
    ethnicities[key] = sum(ethnicities[key]) / len(ethnicities[key])
# сортируем словарь по алфавиту, чтобы было красивее
ethnicities = dict(sorted(ethnicities.items()))
ethnicities

CPU times: user 31 ms, sys: 786 µs, total: 31.8 ms
Wall time: 59.7 ms


{'group A': 64.67415730337079,
 'group B': 67.35263157894737,
 'group C': 69.10344827586206,
 'group D': 70.03053435114504,
 'group E': 73.02857142857142}

Время выполнения больше почти в 10 раз (3.24 и 31 ms)

Для каждого уровня образования выведите минимальный балл за экзамен по письму

In [ ]:
# чтобы засечь время
%%time
# создаем словарь с уровнями образования, куда будем записывать баллы по письму
educations = {}
# для каждой строчки в таблице
for i in range(len(df['education'])):
    # если уровень образования еще не встречался в словаре
    if df['education'][i] not in educations:
        # создаем для этого уровня образования словарь, где будут баллы по письму
        educations[df['education'][i]] = []
        # добавляем балл по письму
        educations[df['education'][i]].append(df['writing score'][i])
    # если уровень образования уже был в словаре
    else:
        # добавляем балл по письму
        educations[df['education'][i]].append(df['writing score'][i])
# для каждого уровня образования в словаре
for key in educations:
    # вместо списка баллов оставляем минимальный
    educations[key] = min(educations[key])
educations = dict(sorted(educations.items()))
educations

CPU times: user 20.7 ms, sys: 0 ns, total: 20.7 ms
Wall time: 20.5 ms


{"associate's degree": 35,
 "bachelor's degree": 38,
 'high school': 15,
 "master's degree": 46,
 'some college': 19,
 'some high school': 10}

Время выполнения больше почти в 4 раза (5.55 b 20.7 ms)


**Задание 13. Выведите на экран средние баллы студентов по каждому предмету в зависимости от пола и уровня образования. То есть должно получиться количество групп, равных 2 * (число уровней образования), и для каждой такой группы выыведите средний балл по каждому из предметов.**

Это можно сделать с помощью сводных таблиц (pivot_table):

https://www.kaggle.com/kamilpolak/tutorial-how-to-use-pivot-table-in-pandas

In [ ]:
# импортируем numpy
import numpy as np

In [ ]:
# создаем своднуб таблицу
# индексы -- это пол и уровень образования
# в знечениях указывается средний балл по каждому из предметов
pd.pivot_table(df, index=["gender", "education"],
               aggfunc={'math score': np.mean,
                        'reading score': np.mean,
                        'writing score': np.mean})

math score  reading score  writing score
gender education                                                   
female associate's degree   65.250000      74.120690      74.000000
       bachelor's degree    68.349206      77.285714      78.380952
       high school          59.351064      68.202128      66.691489
       master's degree      66.500000      76.805556      77.638889
       some college         65.406780      73.550847      74.050847
       some high school     59.296703      69.109890      68.285714
male   associate's degree   70.764151      67.433962      65.405660
       bachelor's degree    70.581818      68.090909      67.654545
       high school          64.705882      61.480392      58.539216
       master's degree      74.826087      73.130435      72.608696
       some college         69.009259      64.990741      63.148148
       some high school     67.840909      64.693182      61.375000

#### Задание 14. Сколько студентов успешно сдали экзамен по математике?

Создайте новый столбец в таблице df под названием Math_PassStatus и запишите в него F, если студент не сдал экзамен по математике (балл за экзамен < passmark), и P иначе.

Посчитайте количество студентов, сдавших и не сдавших экзамен по математике.

Сделайте аналогичные шаги для экзаменов по чтению и письму.

In [ ]:
# создаем столбцы: если балл больше или равен passmark, то пишем Р, иначе -- F
df['Math_PassStatus'] = np.where(df['math score'] >= passmark, 'P', 'F')
df['Reading_PassStatus'] = np.where(df['reading score'] >= passmark, 'P', 'F')
df['Writing_PassStatus'] = np.where(df['writing score'] >= passmark, 'P', 'F')
df

gender race/ethnicity           education         lunch test preparation  \
0    female        group B   bachelor's degree      standard             none   
1    female        group C        some college      standard        completed   
2    female        group B     master's degree      standard             none   
3      male        group A  associate's degree  free/reduced             none   
4      male        group C        some college      standard             none   
..      ...            ...                 ...           ...              ...   
995  female        group E     master's degree      standard        completed   
996    male        group C         high school  free/reduced             none   
997  female        group C         high school  free/reduced        completed   
998  female        group D        some college      standard        completed   
999  female        group D        some college  free/reduced             none   

       group  math score  reading score  writing score Math_PassStatus  \
0     group1          72             72             74               P   
1     group1          69             90             88               P   
2     group1          90             95             93               P   
3     group1          47             57             44               F   
4     group1          76             78             75               P   
..       ...         ...            ...            ...             ...   
995  group10          88             99             95               P   
996  group10          62             55             55               P   
997  group10          59             71             65               P   
998  group10          68             78             77               P   
999  group10          77             86             86               P   

    Reading_PassStatus Writing_PassStatus  
0                    P                  P  
1                    P                  P  
2                    P                  P  
3                    P                  F  
4                    P                  P  
..                 ...                ...  
995                  P                  P  
996                  P                  P  
997                  P                  P  
998                  P                  P  
999                  P                  P  

[1000 rows x 12 columns]

In [ ]:
# количество сдавших и не сдавших по математике
df['Math_PassStatus'].value_counts()

P    865
F    135
Name: Math_PassStatus, dtype: int64

In [ ]:
# количество сдавших и не сдавших по чтению
df['Reading_PassStatus'].value_counts()

P    910
F     90
Name: Reading_PassStatus, dtype: int64

In [ ]:
# количество сдавших и не сдавших по письму
df['Writing_PassStatus'].value_counts()

P    886
F    114
Name: Writing_PassStatus, dtype: int64

#### Задание 15. Сколько студентов успешно сдали все экзамены?

Создайте столбец OverAll_PassStatus и запишите в него для каждого студента 'F', если студент не сдал хотя бы один из трех экзаменов, а иначе 'P'.

Посчитайте количество студентов, которые сдали все экзамены.

In [ ]:
# создаем столбец по трем условиям: если в трех созданных выше столбцах Р, то пишем Р, иначе -- F
df['Overall_PassStatus'] = np.where(((df['Math_PassStatus'] == 'P') &
                                     (df['Reading_PassStatus'] == 'P') &
                                     (df['Writing_PassStatus'] == 'P')),
                                    'P', 'F')
df

gender race/ethnicity           education         lunch test preparation  \
0    female        group B   bachelor's degree      standard             none   
1    female        group C        some college      standard        completed   
2    female        group B     master's degree      standard             none   
3      male        group A  associate's degree  free/reduced             none   
4      male        group C        some college      standard             none   
..      ...            ...                 ...           ...              ...   
995  female        group E     master's degree      standard        completed   
996    male        group C         high school  free/reduced             none   
997  female        group C         high school  free/reduced        completed   
998  female        group D        some college      standard        completed   
999  female        group D        some college  free/reduced             none   

       group  math score  reading score  writing score Math_PassStatus  \
0     group1          72             72             74               P   
1     group1          69             90             88               P   
2     group1          90             95             93               P   
3     group1          47             57             44               F   
4     group1          76             78             75               P   
..       ...         ...            ...            ...             ...   
995  group10          88             99             95               P   
996  group10          62             55             55               P   
997  group10          59             71             65               P   
998  group10          68             78             77               P   
999  group10          77             86             86               P   

    Reading_PassStatus Writing_PassStatus Overall_PassStatus  
0                    P                  P                  P  
1                    P                  P                  P  
2                    P                  P                  P  
3                    P                  F                  F  
4                    P                  P                  P  
..                 ...                ...                ...  
995                  P                  P                  P  
996                  P                  P                  P  
997                  P                  P                  P  
998                  P                  P                  P  
999                  P                  P                  P  

[1000 rows x 13 columns]

In [ ]:
# количество тех, кто сдал все
len(df[df['Overall_PassStatus'] == 'P'])

812

##### Задание 16. Переведем баллы в оценки

### Система перевода баллов в оценки
####    больше 90 = A
####      80-90 = B
####      70-80 = C
####      60-70 = D
####      50-60 = E
####    меньше 50 = F (Fail)

Создайте вспомогательную функцию, которая будет по среднему баллу за три экзамена выставлять оценку студенту по данным выше критериям.

Создайте столбец Grade и запишите в него оценку каждого студента.

Выведите количество студентов, получивших каждую из оценок.

**В случае, если средний балл попадает на границу между оценками (т.е. равен ровно 60, 70 или 80 баллов), вы можете интерпретировать условие на своё усмотрение (т.е. можете поставить за 60 баллов оценку D, а можете - E).**

In [ ]:
# функция перевода балла в оценку (граничный балл относится к оценке выше)
def GetGrade(average_mark):
    if average_mark >= 90:
        return 'A'
    elif average_mark >= 80:
        return 'B'
    elif average_mark >= 70:
        return 'C'
    elif average_mark >= 60:
        return 'D'
    elif average_mark >= 50:
        return 'E'
    else:
        return 'F'

# создаем новый столбец, применяя функцию GetGrade к каждому элементу, который определяется как
# среднее по трем полученным баллам за экзамены
df['Grade'] = np.array(list(map(GetGrade,
 ((df['math score'] + df['reading score'] + df['writing score']) / 3))))
df

gender race/ethnicity           education         lunch test preparation  \
0    female        group B   bachelor's degree      standard             none   
1    female        group C        some college      standard        completed   
2    female        group B     master's degree      standard             none   
3      male        group A  associate's degree  free/reduced             none   
4      male        group C        some college      standard             none   
..      ...            ...                 ...           ...              ...   
995  female        group E     master's degree      standard        completed   
996    male        group C         high school  free/reduced             none   
997  female        group C         high school  free/reduced        completed   
998  female        group D        some college      standard        completed   
999  female        group D        some college  free/reduced             none   

       group  math score  reading score  writing score Math_PassStatus  \
0     group1          72             72             74               P   
1     group1          69             90             88               P   
2     group1          90             95             93               P   
3     group1          47             57             44               F   
4     group1          76             78             75               P   
..       ...         ...            ...            ...             ...   
995  group10          88             99             95               P   
996  group10          62             55             55               P   
997  group10          59             71             65               P   
998  group10          68             78             77               P   
999  group10          77             86             86               P   

    Reading_PassStatus Writing_PassStatus Overall_PassStatus Grade  
0                    P                  P                  P     C  
1                    P                  P                  P     B  
2                    P                  P                  P     A  
3                    P                  F                  F     F  
4                    P                  P                  P     C  
..                 ...                ...                ...   ...  
995                  P                  P                  P     A  
996                  P                  P                  P     E  
997                  P                  P                  P     D  
998                  P                  P                  P     C  
999                  P                  P                  P     B  

[1000 rows x 14 columns]